src/xmipp/libraries/reconstruction/resolution_fso.cpp

```diff
@@ -0,0 +1,1877 @@
+
+            		f1_mean += cosine*conj(DIRECT_MULTIDIM_ELEM(f1,n));
+            		f2_mean += cosine*conj(DIRECT_MULTIDIM_ELEM(f2,n));
+        		}
+        	}
+        }
+
+        cross += abs(f1_orig - f2_mean)*abs(f1_orig - f2_mean) +
+        		abs(f2_orig - f1_mean)*abs(f2_orig - f1_mean);
+    }
+
+
+    void fscShell(MultidimArray< std::complex< double > > & FT1,
+    		 MultidimArray< std::complex< double > > & FT2,
+			 Matrix1D<double> &freq_fourier_x,
+			 Matrix1D<double> &freq_fourier_y,
+			 Matrix1D<double> &freq_fourier_z,
+			 MultidimArray< double >& freqMap,
+			 int m1sizeX, Matrix2D<double> &indexesFourier, Matrix2D<double> &indexesFourier2, double &cutoff,
+			 MultidimArray<std::complex<double>> &f1, MultidimArray<std::complex<double>> &f2)
+    {
+    	int idxcutoff = (int) round(cutoff * m1sizeX);
+        int ZdimFT1=(int)ZSIZE(FT1);
+        int YdimFT1=(int)YSIZE(FT1);
+        int XdimFT1=(int)XSIZE(FT1);
+
+        long n = 0;
+        int Nelems = 0;
+        for (int k=0; k<ZdimFT1; k++)
+        {
+            for (int i=0; i<YdimFT1; i++)
+            {
+                for (int j=0; j<XdimFT1; j++)
+                {
+
+                    double f = 1/DIRECT_MULTIDIM_ELEM(freqMap,n);
+
+                    if (DIRECT_MULTIDIM_ELEM(freqMap,n)<2)
+                    {
+                    	++n;
+                    	continue;
+                    }
+
+                    ++n;
+                    int idx = (int) round(f * m1sizeX);
+
+                    if (idx != idxcutoff)
+                    	continue;
+                    Nelems++;
+                }
+            }
+        }
+
+        f1.initZeros(Nelems);
+        f2.initZeros(Nelems);
+        int counter = 0;
+
+        indexesFourier.initZeros(3,Nelems);
+        indexesFourier2 = indexesFourier;
+
+        n = 0;
+        for (int k=0; k<ZdimFT1; k++)
+		{
+			for (int i=0; i<YdimFT1; i++)
+			{
+				for (int j=0; j<XdimFT1; j++)
+				{
+					double iu = DIRECT_MULTIDIM_ELEM(freqMap,n);
+
+                    if (iu<2)
+                    {
+                    	++n;
+                    	continue;
+                    }
+                    double f = 1/iu;
+					++n;
+					int idx = (int) round(f * m1sizeX);
+
+					if (idx != idxcutoff)
+						continue;
+
+					DIRECT_MULTIDIM_ELEM(f1,counter) = dAkij(FT1, k, i, j);
+					DIRECT_MULTIDIM_ELEM(f2,counter) = dAkij(FT2, k, i, j);
+
+					MAT_ELEM(indexesFourier, 0, counter) = VEC_ELEM(freq_fourier_z,k)*iu;
+					MAT_ELEM(indexesFourier, 1, counter) = VEC_ELEM(freq_fourier_x,j)*iu;
+					MAT_ELEM(indexesFourier, 2, counter) = VEC_ELEM(freq_fourier_y,i)*iu;
+//					MAT_ELEM(indexesFourier, 3, counter) = iu;
+					MAT_ELEM(indexesFourier2, 0, counter) = k;
+					MAT_ELEM(indexesFourier2, 1, counter) = j;
+					MAT_ELEM(indexesFourier2, 2, counter) = i;
+					counter++;
+				}
+			}
+		}
+    }
+
+
+    void generateDirections(Matrix2D<double> &angles, bool alot)
```

Comment:
a lot of directions to avoid implement a method that generates them. Picking points on the sphere is not a so simple problem as it seems, and there are multiple solutions.